event/filters.py

```diff
@@ -0,0 +1,120 @@
+from .models import Event, Like
+from user.models import OcialUser
+from ocial.models import Category
+from datetime import datetime
+from rest_framework.authtoken.models import Token
+from django.db.models.functions import ACos, Cos, Radians, Sin
+from django.db.models import F
+
+
+def filterEvents(request):
+    params = request.query_params
+    filters = params.keys() 
+    if ("timeStart" in filters and "timeEnd" in filters):
+        if params["timeStart"] > params["timeEnd"]:
+            return {
+                "error": "Invalid time",
+                "message": "timeStart must be before timeEnd",
+            }
+    if not (("longitude" in filters and "latitude" in filters and "radius" in filters) or (
+        "longitude" not in filters and "latitude" not in filters and "radius" not in filters
+    )):
+        return {
+            "error": "Invalid location",
+            "message": "You must provide all of the location filter parameters: longitude, latitude, radius",
+        }
+    events = Event.objects.filter(timeEnd__gte=datetime.now())
+    for param in filters:
+        events = filter(param, params[param], events, request)
+    if "longitude" in filters:
+        return filterLocation(
+            float(params["longitude"]),
+            float(params["latitude"]),
+            float(params["radius"]),
+            events
+        )
+    return events
+
+
+def filter(param, value, events, request):
+    match param:
+        case "all":
+            if value == "true":
+                return Event.objects.all()
+            else:
+                return events
+        case "category":
+            if value not in [
+                Category.SPORTS.value,
+                Category.MUSIC.value,
+                Category.MARKETS.value,
+                Category.RELAX_ACTIVITIES.value,
+                Category.LIVE_CONCERT.value,
+            ]:
+                return {
+                    "error": "Invalid category",
+                    "message": "Category must be one of the following: Sports, Music, Markets, Relax activities, Live concert",
+                }
+            return events.filter(category=value)
+        case "timeStart":
+            return events.filter(timeStart__gte=value)
+        case "timeEnd":
+            return events.filter(timeStart__lte=value)
+        case "liked":
```

Comment:
Esto hace que los filtros sean exclusivos, en vez de inclusivos. Es decirr, solo podemos filtrar por un parámetro a la vez, no por todos o algunos.

Reply:
Lo que yo le dije a Carlos q podéis hacer es meter cada una de las funciones en un array en el ``get`` del controlador y en un bucle ejecutarlas y concatenar sus resultados en un Set (para evitar duplicados).

Reply:
`filter` es para poner cada filtro por separado. Si te fijas, en `filterEvents` recoje aquellos filtros que haya seleccionado el usuario, ya que obtiene qué `params` se han seleccionado y ejecuta el filtrado inclusivo. He comprobado la implementación y funciona el filtrado por varios filtros.

Reply:
@carvarsou Pero [aquí](https://github.com/ispp-2324-ocial/backend/blob/d16811e8b1460073e324a5ad570d2bfa9c628bc1/event/filters.py#L28) estamos sobreescribiendo en cada bucle la variable ``events``, en vez de concatenarlo con la anterior, no?

Si filtras por like y por Categoría, te sale bien?

Reply:
Adjunto capturas donde se ve la diferencia entre los dos eventos de categoría "Markets", siendo uno de ellos el que tiene like.

<img src="https://github.com/ispp-2324-ocial/backend/assets/80246400/9e0fd3e4-2f49-4d85-b12e-cd224aac6b4a" width="320" height="320">
<img src="https://github.com/ispp-2324-ocial/backend/assets/80246400/d6e349cd-cfbf-4d6c-8fee-6eeeceba6e37" width="320" height="320">

En cuanto a tu primera pregunta, es cierto que por cada parámetro se sobrescribe la variable `events`. Entonces sería como has dicho antes para evitar esto:

> Lo que yo le dije a Carlos q podéis hacer es meter cada una de las funciones en un array en el `get` del controlador y en un bucle ejecutarlas y concatenar sus resultados en un Set (para evitar duplicados).

Reply:
events es un queryset de django, no se sobre escribe sino que se va aplicando cada filtro de manera sucesiva